tests/functional_tests/team_barrier_tester.cpp

```diff
@@ -0,0 +1,115 @@
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ *****************************************************************************/
+
+
+rocshmem_team_t team_barrier_world_dup;
+
+/******************************************************************************
+ * Device TEST KERNEL
+ *****************************************************************************/
+__global__ void TeamBarrierTest(int loop, int skip, long long int *start_time,
+                                  long long int *end_time,
+                                  ShmemContextType ctx_type,
+                                  rocshmem_team_t *teams) {
+  __shared__ rocshmem_ctx_t ctx;
+  int wg_id = get_flat_grid_id();
+
+  rocshmem_wg_init();
+  rocshmem_wg_team_create_ctx(teams[wg_id], ctx_type, &ctx);
+
+  int n_pes = rocshmem_ctx_n_pes(ctx);
+
+  __syncthreads();
+
+  for (int i = 0; i < loop + skip; i++) {
+    if (i == skip && hipThreadIdx_x == 0) {
+      start_time[wg_id] = wall_clock64();
+    }
+
+    rocshmem_barrier(teams[wg_id]);
+  }
+
+  __syncthreads();
+
+  if (hipThreadIdx_x == 0) {
+    end_time[wg_id] = wall_clock64();
+  }
+
+  rocshmem_wg_ctx_destroy(&ctx);
+  rocshmem_wg_finalize();
+}
+
+/******************************************************************************
+ * HOST TESTER CLASS METHODS
+ *****************************************************************************/
+TeamBarrierTester::TeamBarrierTester(TesterArguments args)
+    : Tester(args){
+  my_pe = rocshmem_team_my_pe(ROCSHMEM_TEAM_WORLD);
+  n_pes = rocshmem_team_n_pes(ROCSHMEM_TEAM_WORLD);
+
+  char* value{nullptr};
+  if ((value = getenv("ROCSHMEM_MAX_NUM_TEAMS"))) {
+    num_teams = atoi(value);
+  }
+
+  CHECK_HIP(hipMalloc(&team_barrier_world_dup,
+                      sizeof(rocshmem_team_t) * num_teams));
+}
+
+TeamBarrierTester::~TeamBarrierTester() {
+  CHECK_HIP(hipFree(team_barrier_world_dup));
+}
+
+void TeamBarrierTester::preLaunchKernel() {
+  for (int team_i = 0; team_i < num_teams; team_i++) {
+    team_barrier_world_dup[team_i] = ROCSHMEM_TEAM_INVALID;
+    rocshmem_team_split_strided(ROCSHMEM_TEAM_WORLD, 0, 1, n_pes, nullptr, 0,
```

Comment:
Would it make sense to test non-contiguous, non-world analogous teams, e.g. `0, 2, n_pes/2` and `1, 2, n_pes/2` (adjusted for n_pes parity)

Reply:
this would definitely make sense, I would suggest to add that as a follow-up item, since this should probably be done for more than just this test.